addons/udes_stock/tests/test_move_line.py

```diff
@@ -0,0 +1,108 @@
+        cls._policy     = cls.picking_type_putaway.u_drop_location_policy
+
+    def tearDown(self):
+        super(TestValidateLocationDest, self).tearDown()
+        self.picking_type_putaway.u_drop_location_constraint = self._constraint
+        self.picking_type_putaway.u_drop_location_policy     = self._policy
+
+    def test01_drop_location_not_suggeested_enforced_failure(self):
+        """ When the policy is `enforce`, an error is thrown when a
+            valid, but not suggested, drop off location is used.
+        """
+        self.picking_type_putaway.u_drop_location_constraint = 'enforce'
+        self.picking_type_putaway.u_drop_location_policy     = 'by_products'
+
+        self.create_quant(self.apple.id, self.test_location_01.id, 4)
+        self.create_quant(self.apple.id,
+                          self.picking_type_putaway.default_location_src_id.id,
+                          4)
+        picking = self.create_picking(self.picking_type_putaway,
+                                      products_info=self._pick_info,
+                                      confirm=True,
+                                      assign=True)
+        mls = picking.move_line_ids
+        locations = picking.get_suggested_locations(mls)
+
+        # We'll use loc 02 to drop off, so we check the assumption
+        self.assertEqual(locations, self.test_location_01)
+        err_msg = "Drop off location must be one of the suggested locations"
+
+        for ml in mls:
+            e = None
+
+            with self.assertRaises(ValidationError) as e:
+                ml.write({'location_dest_id': self.test_location_02.id})
+
+            self.assertEqual(e.exception.name, err_msg)
+
+    def test02_drop_location_not_suggeested_not_enforced(self):
+        """ When the policy is `suggest`, NO error is thrown when a
+            valid, but not suggested, drop off location is used.
+            NB: same as test01, but with 'suggest' constraint.
+        """
+        self.picking_type_putaway.u_drop_location_constraint = 'suggest'
+        self.picking_type_putaway.u_drop_location_policy     = 'by_products'
+
+        self.create_quant(self.apple.id, self.test_location_01.id, 4)
+
+        self.create_quant(self.apple.id,
+                          self.picking_type_putaway.default_location_src_id.id,
+                          4)
+        picking = self.create_picking(self.picking_type_putaway,
+                                      products_info=self._pick_info,
+                                      confirm=True,
+                                      assign=True)
+        mls = picking.move_line_ids
+        locations = picking.get_suggested_locations(mls)
+
+        # We'll use loc 02 to drop off, so we check the assumption
+        self.assertEqual(locations, self.test_location_01)
+
+        for ml in mls:
+            ml.write({'location_dest_id': self.test_location_02.id})
+            self.assertTrue(True)
```

Comment:
why do this? it literally does nothing

Reply:
It tells you're testing.

Reply:
It by very definition can never fail. I'm missing something here :confused:

Reply:
Code is meant to be read. Don't miss that.